### PR TITLE
Add 3 endpoints in Patients API

### PR DESCRIPTION
**Pull request by ersinciftci**
POST /patients/fetch
GET /studies/{studyId}/patients
GET /studies/{studyId}/patients/{patientId}

Backend development guideline (draft):
https://docs.google.com/document/d/1HjOXWMq4z1JjeZLkw4eq3aG_o0y3IhZa47bGUtdoJ3Y

### CODE REVIEW

File: persistence/persistence-mybatis-test/src/test/java/org/cbioportal/persistence/mybatis/PatientMyBatisRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    public void fetchPatients() throws Exception {
+
+        List<String> studyIds = new ArrayList<>();
+        studyIds.add("study_tcga_pub");
+        studyIds.add("study_tcga_pub");
+        List<String> patientIds = new ArrayList<>();
+        patientIds.add("TCGA-A1-A0SB");
+        patientIds.add("TCGA-A1-A0SE");
+
+        List<Patient> result = patientMyBatisRepository.fetchPatients(studyIds, patientIds, "SUMMARY");
```
Comment (sheridancbio) [inline persistence/persistence-mybatis-test/src/test/java/org/cbioportal/persistence/mybatis/PatientMyBatisRepositoryTest.java]: This pairing of arguments by array index feels wrong .. but I guess doing it this way allows mapping to the same handler in the mybatis mapper, rather than having one handler for a list of (study,patient) pairing objects .. and a different handler for a simple list of study identifiers (patientIds=null).

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/PatientMapper.xml [hunk 2]
```diff
@@ -19,4 +19,57 @@
...
+
+    <sql id="from">
+        FROM patient
+        INNER JOIN cancer_study ON patient.CANCER_STUDY_ID = cancer_study.CANCER_STUDY_ID
+    </sql>
+
+    <sql id="where">
+        <where>
+            <if test="patientIds == null">
+                cancer_study.CANCER_STUDY_IDENTIFIER = #{studyIds[0]}
```
Comment (sheridancbio) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/PatientMapper.xml]: so .. without patientIds, we will only look at the first study identifier in the list?  Intuitively, if a caller to the repository passes in a list of cancer studies, wouldn't they expect the union of all patients from all of the studies passed in?

perhaps the repository should throw an exception when called with a list of studies, but no patient Ids. (or at least, we should put comments by this repository functions in the interface explaining that such a call will ignore all but the first study)

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/PatientMapper.xml [hunk 3]
```diff
@@ -19,4 +19,57 @@
...
+        SELECT
+        <include refid="select">
+            <property name="prefix" value=""/>
+        </include>
+        <include refid="from"/>
+        <include refid="where"/>
+        <if test="sortBy != null and projection != 'ID'">
+            ORDER BY ${sortBy} ${direction}
+        </if>
+        <if test="projection == 'ID'">
```
Comment (sheridancbio) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/PatientMapper.xml]: note to self: we do order by STABLE_ID on an ID projection here. (hardcode order ASC)

File: service/src/main/java/org/cbioportal/service/exception/PatientNotFoundException.java [hunk 4]
```diff
@@ -2,13 +2,23 @@
 
 public class PatientNotFoundException extends Exception {
```
Comment (sheridancbio) [inline service/src/main/java/org/cbioportal/service/exception/PatientNotFoundException.java]: this now functions as "Patient Not Found In Study" exception (rather than global search) .. or maybe it handles both cases.

File: web/src/main/java/org/cbioportal/web/PatientController.java [hunk 5]
```diff
@@ -14,45 +21,86 @@
...
-    @RequestMapping(value = "/studies/{studyId}/patients/{patientId}", method = RequestMethod.GET)
+    @RequestMapping(value = "/studies/{studyId}/patients/{patientId}", method = RequestMethod.GET,
+        produces = MediaType.APPLICATION_JSON_VALUE)
     @ApiOperation("Get a patient in a study")
-    public ResponseEntity<Patient> getPatientInStudy(@PathVariable String studyId,
-                                                            @PathVariable String patientId) {
+    public ResponseEntity<Patient> getPatientInStudy(
+        @ApiParam(required = true, value = "Study ID e.g. acc_tcga")
+        @PathVariable String studyId,
+        @ApiParam(required = true, value = "Patient ID e.g. TCGA-OR-A5J2")
```
Comment (sheridancbio) [inline web/src/main/java/org/cbioportal/web/PatientController.java]: Good guidance with examples

File: web/src/main/java/org/cbioportal/web/PatientController.java [hunk 6]
```diff
@@ -14,45 +21,86 @@
...
+        @ApiParam(required = true, value = "List of patient identifiers")
+        @RequestBody List<PatientIdentifier> patientIdentifiers) throws PageSizeTooBigException {
 
-        throw new UnsupportedOperationException();
-    }
+        if (patientIdentifiers.size() > PagingConstants.MAX_PAGE_SIZE) {
+            throw new PageSizeTooBigException(patientIdentifiers.size());
+        }
+
+        List<String> studyIds = new ArrayList<>();
```
Comment (sheridancbio) [inline web/src/main/java/org/cbioportal/web/PatientController.java]: argument processing inside controller : ok
the list construction code is reasonably short ; when the logic gets heavier I would start suggesting a util class

File: web/src/main/java/org/cbioportal/web/PatientController.java [hunk 7]
```diff
@@ -14,45 +21,86 @@
...
+        List<String> patientIds = new ArrayList<>();
 
-    @RequestMapping(value = "/patients/query", method = RequestMethod.POST)
-    @ApiOperation("Query patients by example")
-    public ResponseEntity<List<Patient>> queryPatientsByExample(@RequestParam(defaultValue = "SUMMARY") Projection projection,
-                                                                                 @RequestParam(defaultValue = PagingConstants.DEFAULT_PAGE_SIZE) Integer pageSize,
-                                                                                 @RequestParam(defaultValue = PagingConstants.DEFAULT_PAGE_NUMBER) Integer pageNumber,
-                                                                                 @RequestBody Patient examplePatient) {
+        for (PatientIdentifier patientIdentifier : patientIdentifiers) {
```
Comment (sheridancbio) [inline web/src/main/java/org/cbioportal/web/PatientController.java]: I see .. this is where the "parallel array" comes from .. the actual argument through the api *is* a pairing of (study,patient)

File: persistence/persistence-mybatis-test/src/test/java/org/cbioportal/persistence/mybatis/PatientMyBatisRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    private PatientMyBatisRepository patientMyBatisRepository;
+
+    @Test
+    public void getAllPatientsInStudyIdProjection() throws Exception {
+
+        List<Patient> result = patientMyBatisRepository.getAllPatientsInStudy("study_tcga_pub", "ID", null, null, null,
+            null);
+
+        Assert.assertEquals(14, result.size());
+        Patient patient = result.get(0);
```
Comment (ersinciftci) [inline persistence/persistence-mybatis-test/src/test/java/org/cbioportal/persistence/mybatis/PatientMyBatisRepositoryTest.java]: This was my reply https://github.com/cBioPortal/cbioportal/pull/1825#discussion_r86157639

File: persistence/persistence-mybatis-test/src/test/java/org/cbioportal/persistence/mybatis/PatientMyBatisRepositoryTest.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        List<Patient> result = patientMyBatisRepository.getAllPatientsInStudy("study_tcga_pub", "SUMMARY", 1, 0, null,
+            null);
+
+        Assert.assertEquals(1, result.size());
+    }
+
+    @Test
+    public void getAllPatientsInStudySummaryProjectionStableIdSort() throws Exception {
+
+        List<Patient> result = patientMyBatisRepository.getAllPatientsInStudy("study_tcga_pub", "SUMMARY", null, null,
```
Comment (ersinciftci) [inline persistence/persistence-mybatis-test/src/test/java/org/cbioportal/persistence/mybatis/PatientMyBatisRepositoryTest.java]: I'll add more test cases eventually (more negative and non-default scenarios).

File: web/src/main/java/org/cbioportal/web/parameter/sort/PatientSortBy.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package org.cbioportal.web.parameter.sort;
+
+public enum PatientSortBy {
+
+    stableId,
```
Comment (sheridancbio) [inline web/src/main/java/org/cbioportal/web/parameter/sort/PatientSortBy.java]: I think it is convention to have enum constants be all upper case. But maybe for the framework to work correctly they must match our data member names.

File: persistence/persistence-mybatis-test/src/test/java/org/cbioportal/persistence/mybatis/PatientMyBatisRepositoryTest.java (hunk 1 above)
Comment (ersinciftci) [inline persistence/persistence-mybatis-test/src/test/java/org/cbioportal/persistence/mybatis/PatientMyBatisRepositoryTest.java]: Yes, it allows us to reuse the queries, and we also get the same pairings from the web layer from fetch endpoint, so makes things easier.

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/PatientMapper.xml (hunk 2 above)
Comment (ersinciftci) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/PatientMapper.xml]: Yes it should throw an error, there should be lots of other error checks, but currently I implement it only for the web layer inputs, since persistence module it's not going to be used by other 3rd party software in a foreseeable future. So my priority is to release these basic APIs in a timely manner, instead of making them impossible to break for 3rd party use. So I say we better leave it for later as well.

File: web/src/test/java/org/cbioportal/web/PatientControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+    public void setUp() throws Exception {
+
+        Mockito.reset(patientService);
+        mockMvc = MockMvcBuilders.webAppContextSetup(wac).build();
+    }
+    
+    @Test
+    public void getAllPatientsInStudyDefaultProjection() throws Exception {
+
+        List<Patient> patientList = createExamplePatients();
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/web/PatientControllerTest.java]: patientList has 2 patients in return value

File: web/src/test/java/org/cbioportal/web/PatientControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+        mockMvc.perform(MockMvcRequestBuilders.get("/studies/test_study_id/patients")
+            .param("projection", "META"))
+            .andExpect(MockMvcResultMatchers.status().isOk())
+            .andExpect(MockMvcResultMatchers.header().string(HeaderKeyConstants.TOTAL_COUNT, "2"));
+    }
+
+    @Test
+    public void getPatientInStudyNotFound() throws Exception {
+
+        Mockito.when(patientService.getPatientInStudy(Mockito.anyString(), Mockito.anyString())).thenThrow(
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/web/PatientControllerTest.java]: test of global handler output format

File: service/src/main/java/org/cbioportal/service/exception/PatientNotFoundException.java (hunk 4 above)
Comment (ersinciftci) [inline service/src/main/java/org/cbioportal/service/exception/PatientNotFoundException.java]: In the current system, Patient id is only meaningful in a Study, so 2 different studies can have Patient id `15` referring to different patients. So I realized that a Patient id without Study id doesn't mean anything, so I added Study id in the exception. Because whenever we query a Patient from database, we do so with Study id.

File: web/src/test/java/org/cbioportal/web/PatientControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+        CancerStudy cancerStudy = new CancerStudy();
+        cancerStudy.setCancerStudyId(TEST_CANCER_STUDY_ID_1);
+        cancerStudy.setCancerStudyIdentifier(TEST_CANCER_STUDY_IDENTIFIER_1);
+        cancerStudy.setTypeOfCancerId(TEST_TYPE_OF_CANCER_ID_1);
+        cancerStudy.setName(TEST_NAME_1);
+        cancerStudy.setShortName(TEST_SHORT_NAME_1);
+        cancerStudy.setDescription(TEST_DESCRIPTION_1);
+        patient.setCancerStudy(cancerStudy);
+
+        Mockito.when(patientService.getPatientInStudy(Mockito.anyString(), Mockito.anyString())).thenReturn(patient);
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/web/PatientControllerTest.java]: one patient return value .. construction could also be externalized to be side-by-side with the two patient constructor (encourage reuse) but it is fine here too

File: web/src/test/java/org/cbioportal/web/PatientControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+        mockMvc.perform(MockMvcRequestBuilders.get("/studies/test_study_id/patients/test_patient_id")
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/web/PatientControllerTest.java]: single patient fetch test

File: web/src/test/java/org/cbioportal/web/PatientControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+    public void fetchPatientsDefaultProjection() throws Exception {
+
+        List<Patient> patientList = createExamplePatients();
+
+        Mockito.when(patientService.fetchPatients(Mockito.anyListOf(String.class), Mockito.anyListOf(String.class),
+            Mockito.anyString())).thenReturn(patientList);
+
+        List<PatientIdentifier> patientIdentifiers = new ArrayList<>();
+        PatientIdentifier patientIdentifier1 = new PatientIdentifier();
+        patientIdentifier1.setStudyId(TEST_CANCER_STUDY_IDENTIFIER_1);
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/web/PatientControllerTest.java]: construction of the patientIdentifiers argument (study,patient) pairs.  Construction could also be externalized (previous comment)

File: web/src/test/java/org/cbioportal/web/PatientControllerTest.java [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+        patientIdentifiers.add(patientIdentifier1);
+        PatientIdentifier patientIdentifier2 = new PatientIdentifier();
+        patientIdentifier2.setStudyId(TEST_CANCER_STUDY_IDENTIFIER_2);
+        patientIdentifier2.setPatientId(TEST_STABLE_ID_2);
+        patientIdentifiers.add(patientIdentifier2);
+
+        mockMvc.perform(MockMvcRequestBuilders.post("/patients/fetch")
+            .accept(MediaType.APPLICATION_JSON)
+            .contentType(MediaType.APPLICATION_JSON)
+            .content(objectMapper.writeValueAsString(patientIdentifiers)))
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/web/PatientControllerTest.java]: pass of constructed parameter into POST request body

File: web/src/test/java/org/cbioportal/web/PatientControllerTest.java [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+            .andExpect(MockMvcResultMatchers.status().isOk())
+            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
+            .andExpect(MockMvcResultMatchers.jsonPath("$", Matchers.hasSize(2)))
+            .andExpect(MockMvcResultMatchers.jsonPath("$[0].internalId").doesNotExist())
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/web/PatientControllerTest.java]: good .. checks that internal ids don't leak out later (@JsonIgnore)

File: web/src/test/java/org/cbioportal/web/PatientControllerTest.java [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+    @Test
+    public void fetchPatientsMetaProjection() throws Exception {
+
+        BaseMeta baseMeta = new BaseMeta();
+        baseMeta.setTotalCount(2);
+
+        Mockito.when(patientService.fetchMetaPatients(Mockito.anyListOf(String.class),
+            Mockito.anyListOf(String.class))).thenReturn(baseMeta);
+
+        List<PatientIdentifier> patientIdentifiers = new ArrayList<>();
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/web/PatientControllerTest.java]: here is reuse opportunity for construction code .. why not externalize it? (not necessary .. up to you)

**Comment by ersinciftci** [comment]
I think it's ready to be merged.
